### ISSUE TITLE
On-demand broker threads and connections

### ISSUE THREAD

**Issue by edenhill** [edenhill/librdkafka#825]
This is a coalescing issue for the following common issues:
 * Sparse connections - just connect to the brokers we need to talk to
 * Sparse threads - just create threads for brokers we need to talk to
 * Purge old brokers - remove brokers no longer reported in metadata

The current idea to solve sparseness is:
Let the broker thread creation / destruction depend on:
 * rkb_toppar_cnt > 0 assigned
 * rkb_outbuf cnt > 0

The problem is that both these are populated by the broker (thread) itself, so we might want to spin up the thread if the rkb_ops queue > 0.

For spinning down it should be the reverse scenario (no toppars, no outbufs, no ops) plus a cooling off time (to avoid thread creation churn), such as a minute or two.

**Comment by Ishiihara** [comment]
@edenhill Do you plan to work on this soon?

**Comment by edenhill** [comment]
@Ishiihara It is on the backlog and work will start after the next release (june).

**Comment by gduranceau** [comment]
@edenhill
Hi Magnus,

I'm pleased to see improvements planned in this area. I was wondering if you considered going even further and use a single reactor thread to handle events from all the broker sockets? In environments with very large Kafka clusters where clients (especially producers) may target all partitions of the topics defined, that would avoid spawning dozens of librdkafka threads in each client.

Also, do you plan to apply these changes for bootstrap broker threads/connections as well?

**Comment by edenhill** [comment]
@gduranceau 
The idea, which is not set in stone, is to use a single IO thread (or possibly one for read and one for write, depending on the facilities the OS provides) that will be highly host-OS specific, to be as optimized as possible.
Then there'll be N worker threads doing the actual parallel-friendly heavy work, batching, compression, etc.

The number of worker threads will be a combination of on-demand and hard configuration limits and depends on the number of partitions (level of parallelism) and the throughput load.

SSL is an unknown here, can we maintain performance using SSL with just a single IO thread that will have to do all the SSL shuffling too?

**Comment by gduranceau** [comment]
@endenhill
Very good. I like the design very much.

I'm not really worried with SSL. I worked on an enterprise service bus serving 300k tps, and SSL was never a concern: symmetric encryption is pretty fast.

One suggestion for the I/O thread (coming from that experience as well): if a single thread is a bottleneck, you can actually support spawning multiple I/O threads (number based on configuration). The key is to make sure to assign each connection to one of these threads only, for all its lifetime, and for all its operations, so as to avoid any sort of concurrency issue. So basically, when a connection is created, one of the I/O threads is chosen with round robin policy, and it remains associated to the connection until the connection terminates.

**Comment by billygout** [comment]
@gduranceau Is this ready and tested? I *think* I really need this guy, for too many open `fd`'s on the kafka broker side.

**Comment by gduranceau** [comment]
@billygout The patch is complete, but so far, this has not been integrated in any production running application on my side yet.

**Comment by billygout** [comment]
@gduranceau thanks! You might just have me as a prod tester in a Kafka producer application.

**Comment by gduranceau** [comment]
@billygout Cool! Don't hesitate to report problems. I will also have to deploy this in a producer application in the weeks to come.

**Comment by jbvmio** [comment]
@gduranceau @billygout How has testing been going with this? (connect.on.demand)

**Comment by billygout** [comment]
I have not tried this yet. Sorry.

**Comment by gduranceau** [comment]
I tested this in a variety of scenarios, but not in a producer application just yet. For info, there's at least one limitation that I'm aware of with my implementation, which I commented  in PR https://github.com/edenhill/librdkafka/pull/1659 (on May 15).

**Comment by chaloulo** [comment]
+1 for this enhancement.

**Comment by amitt001** [comment]
@edenhill I see that enable.sparse.connections flag is added to librdkafka. Are these two features present in the latest version and which is the latest version with these options?
* Sparse connections - just connect to the brokers we need to talk to
* Sparse threads - just create threads for brokers we need to talk to

Also, I am wondering how is it going to behave for a Producer. The "Introduction" docs say that:

"Random broker selection
When there is no broker connection and a connection to any broker is needed, such as on startup to retrieve metadata, the client randomly selects a broker from its list of brokers"

My assumption is that initially when a producer is configured it's not producing to any topic so it connects to all the brokers. I am seeing the same behavior in my test. As soon as I am configuring the producer with two brokers I see 6 new threads. But the thread count is not going down even when producing to a topic.

**Comment by edenhill** [comment]
@amitt001 The next version of librdkafka, v1.0.0, will feature sparse connections, it will only connect to the brokers it needs to communicate with. But there will still be one thread per known broker, i.e., no sparse threads at this point, that requires a larger re-design of the librdkafka internals and is not scheduled for the short term.
The non-connected threads will be mostly idle though.

**Comment by fuyijun1989** [comment]
@edenhill My team has a scenario that dozens of clients run in one machine and each connects to 300+ brokers, and due to the 1 thread to 1 brokers design, the thread number goes up to 10K, resulting in severe cpu burn on context switch. If we won't get a "sparse thread" config in the near future, could you give us some advices to mitigate the situation? Maybe some minor code chage or config change.
Many thanks.

**Comment by edenhill** [comment]
We're preparing our Q1 backlog and I'm doing my best to get sparse threads on there.
The idea is to only spin up a broker's thread when it is actually needed (to housekeep partitions or communicate with a broker), and then kill them after some idle period again. This should remedy the situation for when a producer or consumer only needs to communicate with a subset of brokers in the cluster, which is usually the case. E.g., the partition leaders for partitions being fetched or produced to.

**Comment by umashankartm** [comment]
@edenhill, we are having an application that connects to multiple Kafka clusters. we are seeing 1000s of threads being created. Do you have any timeline on when can we expect this feature? 
Thanks you..

**Comment by billygout** [comment]
@umashankartm the following PR was already merged years ago: https://github.com/edenhill/librdkafka/pull/2019

You'll find it referenced in release notes from back then with accompanying config param documentation.

If you are using that already, and still are lots of connections, perhaps your Kafka cluster is 1000's of nodes large, and you are consuming from or producing to at least one partition led by each broker? Otherwise, no idea.

**Comment by umashankartm** [comment]
@billygout thanks for your quick response.. my understanding was that only sparse connection was merged not sparse threads. Sparse connection does help to some extent but issue we are facing right now is with respect to the number of OS threads. Moreover, i dont see any documentation related sparse threads in the latest code base. Am I missing something?

**Comment by billygout** [comment]
@umashankartm good point. I don't remember whether the threads were cut down along with the connections, but even before that change, I seem to recall the number of threads being one per broker, plus a small number of special purpose threads. Also, if you're running on Linux, not that the librdkafka uses the thread naming feature, and it may help diagnose if you can show the thread names, by, for example, running top In thread more for your process PID with `top -H -p <PID>` by hitting the `c` key, which should toggle the display of the COMMAND column to show thread names.

**Comment by bobhanckel** [comment]
Is the "too many threads" issue still a work in progress?   If so is there a ball park figure as to when this feature will become available.   Thanks.

**Comment by edenhill** [comment]
Nothing committed, but we might address this in Q3 or Q4.
It does seem to be less of an issue lately though.

**Comment by bobhanckel** [comment]
Thanks Magnus.   Appreciate your help and feedback.

Q3 or Q4 might work for us to get it in for the next major
release of Oracle.

________________________________

**Comment by mtrienis** [comment]
@edenhill 

> It does seem to be less of an issue lately though.

We have a use case where we need to spin up multiple instances of confluent-kafka-python to split out traffic to different Kafka clusters. Sparse threads would be super helpful here as we are trying to minimize operational complexity / overhead of too many threads for services that integrate this library.

**Comment by edenhill** [comment]
@mtrienis Can you elaborate on the operational complexity that comes with the current threading model?

**Comment by mtrienis** [comment]
@edenhill Most of the Python application I work with are single threaded, so when we integrate those apps with confluent-kafka-python we're adding roughly N + 1 threads where N is the number of brokers. So if we are publishing data to a 40 node Kafka cluster, we're adding roughly 41 threads to every single Python process across every service. If we want two instances of confluent-kafka-python, then we end up with 82 threads, ect ... That's quite a few threads to add to an application that would otherwise be fairly simple (i.e. single threaded). If that application owner needed to spawn a couple threads for their own business logic (i.e. not data related), then we could imagine it could complicate their debugging process a bit.

**Comment by Suresh-Vaidyanath** [comment]
@edenhill Is there any version that has implemented the enhancement ? If not, is there a plan to implement the same and tentative time lines for the same.

**Comment by abhimanyuseth-alation** [comment]
We have this issue as well. We're using Confluent Kafka and have an application using confluent-kafka-go library. Our Confluent Kafka cluster has 30 brokers, and the app has 8 consumers. So we have 248 threads always running.
In addition, there are cron jobs for various retry queues that run regularly, and we see the number of threads increasing to 600+.

**Comment by zhongyuankai** [comment]
Is the problem of too many threads still continuing?
